Compilation SymuVia Version OpenSymuVia + `SymAlterRouteEx`

Hello 

I'm trying at this point to compile the Symuvia version with the function `symAlterRouteEx` provided by @becarie. I'm going through the recipe to build a package that I'm willing to push in my personal channel and share with Cecile Daniel for her research work. At the moment I'm getting some errors. Please some details below 

#### French version

En ce moment j’essaie de compiler la version de Symuvia avec `symAlterRouteEx`  que tu nous a préparé @becarie. Je passe par `conda-build` tout simplement, car j'essaie de construire un paquet conda que j'ai envie de mettre sur mon canal perso. Tout ca a fin de developer les activités avec Cecile Daniel. 

## Conda Build

```
mkdir conda-bld
conda build . -c conda-forge -c licit-lab --output-folder conda-bld/
```
### Error

```
max(const _Tp& __a, const _Tp& __b, _Compare __comp)
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [symumaster/CMakeFiles/SymuMaster.dir/build.make:300: symumaster/CMakeFiles/SymuMaster.dir/Simulation/Simulators/SimulationDescriptor.cpp.o] Error 1
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627397946434/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3072:25: warning: expression result unused [-Wunused-value]
                    for(itUser; itUser!=lstAllUserByOD.end();)
                        ^~~~~~
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627397946434/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3079:33: warning: expression result unused [-Wunused-value]
                            for(itMapValueted; itMapValueted != mapValuetedPath.end(); itMapValueted++)
                                ^~~~~~~~~~~~~
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627397946434/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3188:25: warning: expression result unused [-Wunused-value]
                    for(itMapValueted; itMapValueted != mapValuetedPath.end();)
                        ^~~~~~~~~~~~~
20 errors generated.
make[2]: *** [symumaster/CMakeFiles/SymuMaster.dir/build.make:286: symumaster/CMakeFiles/SymuMaster.dir/Simulation/SimulationRunner.cpp.o] Error 1
3 warnings generated.
make[1]: *** [CMakeFiles/Makefile2:217: symumaster/CMakeFiles/SymuMaster.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
Traceback (most recent call last):
  File "/Users/andresladino/opt/anaconda3/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 481, in main
    execute(sys.argv[1:])
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 472, in execute
    verify=args.verify, variants=args.variants, cache_dir=args.cache_dir)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/api.py", line 195, in build
    variants=variants
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/build.py", line 3073, in build_tree
    notest=notest,
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/build.py", line 2192, in build
    cwd=src_dir, stats=build_stats)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/utils.py", line 392, in _func_defaulting_env_to_os_environ
Merge branch 'dev' of https://github.com/aladinoster/Open-SymuVia into dev
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627397946434/work/conda_build.sh']' returned non-zero exit status 2.
```

## Own build

I've tried:

Preparing the environment
```
conda create -n py39 python=3.9
conda activate py39
conda env update -f https://raw.githubusercontent.com/licit-lab/Open-SymuVia/master/conda/env.yaml
conda install -c conda-forge cmake 
conda install -c anaconda make 
``` 

Compiling does not detect the Xqilla module. 

```
❯ mkdir build;cd build;cmake ..;make -j6
-- The C compiler identification is AppleClang 11.0.3.11030032
-- The CXX compiler identification is AppleClang 11.0.3.11030032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /usr/local/lib/cmake/Boost-1.72.0/BoostConfig.cmake (found version "1.72.0") found components: chrono date_time locale log thread serialization wserialization filesystem system program_options python
-- Library Boost is available
-- 	 Boost::chrono
-- 	 Boost::date_time
-- 	 Boost::locale
-- 	 Boost::serialization
-- 	 Boost::wserialization
-- 	 Boost::python
-- 	 Boost::log
-- 	 Boost::filesystem
-- 	 Boost::thread
--
-- 	 Boost::system
-- 	 Boost headers included
-- Found GDAL: /usr/local/Cellar/gdal/3.1.2_1/lib/libgdal.dylib (found version "3.1.2")
-- Library GDAL is available
-- 	 GDAL library: /usr/local/Cellar/gdal/3.1.2_1/lib/libgdal.dylib,
 	 Headers in: /usr/local/include
-- 	 GDAL headers included
-- Found XercesC: /usr/local/lib/libxerces-c.dylib (found version "3.2.3")
-- Library XercesC is available
-- 	 XercesC library: /usr/local/lib/libxerces-c.dylib,
 	 Headers in: /usr/local/include
-- 	 XercesC headers included
-- Found ODBC: /usr/local/lib/libodbc.dylib
-- Library ODBC is available
-- 	 ODBC library: /usr/local/lib/libodbc.dylib,
 	 Headers in: /usr/local/include
-- 	 ODBC headers included
-- Found Python: /Users/andresladino/opt/anaconda3/envs/py39/bin/python3.9 (found version "3.9.5") found components: Interpreter Development Development.Module Development.Embed
-- Python is available
-- 	 Python library: /Users/andresladino/opt/anaconda3/envs/py39/lib/libpython3.9.dylib,
 	 Headers in: /Users/andresladino/opt/anaconda3/envs/py39/include/python3.9
-- 	 Python headers included
-- RapidJSON found. Headers: /usr/local/Cellar/rapidjson/1.1.0/include
-- Rapidjson is available
CMake Error at cmake/modules/FindXqilla.cmake:5 (find_path):
  Could not find XQILLA_INCLUDE_DIR using the following files: name,
  xqilla-dom3.hpp
Call Stack (most recent call first):
  CMakeLists.txt:162 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/andresladino/Documents/01-Code/02-Python/research/control-routing/symuvia/build/CMakeFiles/CMakeOutput.log".
See also "/Users/andresladino/Documents/01-Code/02-Python/research/control-routing/symuvia/build/CMakeFiles/CMakeError.log".
make: *** No targets specified and no makefile found.  Stop.
```

I've tried this: 

```
mkdir build
cd build
cmake .. -DCMAKE_PREFIX_PATH=$CONDA_PREFIX -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX
make -j6
```

But still I get this: 
```
CMake Error at cmake/modules/FindXqilla.cmake:5 (find_path):
  Could not find XQILLA_INCLUDE_DIR using the following files: name,
  xqilla-dom3.hpp
Call Stack (most recent call first):
  CMakeLists.txt:162 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/andresladino/Documents/01-Code/02-Python/research/control-routing/symuvia/build/CMakeFiles/CMakeOutput.log".
```

@floriangc Ok step 1. solved. In fact xqilla was not being installed within the environment with the `conda update` instruction.
```
conda create -n py39 python=3.9
conda activate py39
conda env update -f https://raw.githubusercontent.com/licit-lab/Open-SymuVia/master/conda/env.yaml
conda install -c conda-forge cmake 
conda install -c anaconda make 
conda install -c licit-lab xqilla
```

I think the error is on compilation rather than the recipe, here is the log of the traces...

```
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:51:14: error: no template named 'map' in namespace 'std'; did you mean 'max'?
        std::map<TraficSimulatorHandler*, std::map<TraficSimulatorHandler*, std::vector<SimulatorsInterface> > > & GetSimulatorsInterfaces();
        ~~~~~^~~
             max
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:2596:1: note: 'max' declared here
max(const _Tp& __a, const _Tp& __b, _Compare __comp)
^
In file included from /Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.cpp:1:
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:51:48: error: no template named 'map' in namespace 'std'; did you mean 'max'?
        std::map<TraficSimulatorHandler*, std::map<TraficSimulatorHandler*, std::vector<SimulatorsInterface> > > & GetSimulatorsInterfaces();
                                          ~~~~~^~~
                                               max
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:2596:1: note: 'max' declared here
max(const _Tp& __a, const _Tp& __b, _Compare __comp)
^
In file included from /Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.cpp:1:
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:51:14: error: non-friend class member 'max' cannot have a qualified name
        std::map<TraficSimulatorHandler*, std::map<TraficSimulatorHandler*, std::vector<SimulatorsInterface> > > & GetSimulatorsInterfaces();
        ~~~~~^
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:51:14: error: C++ requires a type specifier for all declarations
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:51:113: error: expected ';' at end of declaration list
        std::map<TraficSimulatorHandler*, std::map<TraficSimulatorHandler*, std::vector<SimulatorsInterface> > > & GetSimulatorsInterfaces();
                                                                                                                ^
                                                                                                                ;
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Simulators/SimulationDescriptor.h:52:20: error: no template named 'map' in namespace 'std'; did you mean 'max'?
        const std::map<TraficSimulatorHandler*, std::map<TraficSimulatorHandler*, std::vector<SimulatorsInterface> > > & GetSimulatorsInterfaces() const;
              ~~~~~^~~
                   max
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/algorithm:2596:1: note: 'max' declared here
max(const _Tp& __a, const _Tp& __b, _Compare __comp)
^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [symumaster/CMakeFiles/SymuMaster.dir/build.make:300: symumaster/CMakeFiles/SymuMaster.dir/Simulation/Simulators/SimulationDescriptor.cpp.o] Error 1
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3072:25: warning: expression result unused [-Wunused-value]
                    for(itUser; itUser!=lstAllUserByOD.end();)
                        ^~~~~~
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3079:33: warning: expression result unused [-Wunused-value]
                            for(itMapValueted; itMapValueted != mapValuetedPath.end(); itMapValueted++)
                                ^~~~~~~~~~~~~
/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/symumaster/Simulation/Assignment/MSADualLoop.cpp:3188:25: warning: expression result unused [-Wunused-value]
                    for(itMapValueted; itMapValueted != mapValuetedPath.end();)
                        ^~~~~~~~~~~~~
20 errors generated.
make[2]: *** [symumaster/CMakeFiles/SymuMaster.dir/build.make:286: symumaster/CMakeFiles/SymuMaster.dir/Simulation/SimulationRunner.cpp.o] Error 1
3 warnings generated.
make[1]: *** [CMakeFiles/Makefile2:217: symumaster/CMakeFiles/SymuMaster.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
Traceback (most recent call last):
  File "/Users/andresladino/opt/anaconda3/bin/conda-build", line 11, in <module>
    sys.exit(main())
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 481, in main
    execute(sys.argv[1:])
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/cli/main_build.py", line 472, in execute
    verify=args.verify, variants=args.variants, cache_dir=args.cache_dir)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/api.py", line 195, in build
    variants=variants
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/build.py", line 3073, in build_tree
    notest=notest,
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/build.py", line 2192, in build
    cwd=src_dir, stats=build_stats)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/utils.py", line 412, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/Users/andresladino/opt/anaconda3/lib/python3.7/site-packages/conda_build/utils.py", line 392, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/Users/andresladino/opt/anaconda3/conda-bld/symuvia_1627402277092/work/conda_build.sh']' returned non-zero exit status 2.
```

--- a/symumaster/Simulation/Simulators/SimulationDescriptor.h
+++ b/symumaster/Simulation/Simulators/SimulationDescriptor.h
@@ -3,6 +3,8 @@
 #ifndef _SYMUMASTER_SIMULATIONDESCRIPTOR_
 #define _SYMUMASTER_SIMULATIONDESCRIPTOR_
 
+#include <map>
+
 #include "SymuMasterExports.h"
 
 #include "SimulatorsInterface.h"

--- a/symumaster/Simulation/Users/UserEngine.h
+++ b/symumaster/Simulation/Users/UserEngine.h
@@ -3,6 +3,8 @@
 #ifndef _SYMUMASTER_USERENGINE_
 #define _SYMUMASTER_USERENGINE_
 
+#include <map>
+
 #include "SymuMasterExports.h"
 
 #include "TrackedUser.h"

--- a/symumaster/Simulation/Users/UserEngineSnapshot.h
+++ b/symumaster/Simulation/Users/UserEngineSnapshot.h
@@ -3,6 +3,8 @@
 #ifndef _SYMUMASTER_USERENGINE_SNAPSHOT_
 #define _SYMUMASTER_USERENGINE_SNAPSHOT_
 
+#include <map>
+
 #include "SymuMasterExports.h"
 
 #include "TrackedUser.h"

